src/dataflow/src/source/postgres.rs

```diff
@@ -0,0 +1,347 @@
+                scalar_type: PgType::from_oid(row.get(0)).unwrap(),
+                nullable: !row.get::<_, bool>(1),
+            })
+            .collect_vec();
+
+        (rel_id, col_types)
+    }
+
+    /// Creates the replication slot and streams the initial snapshot of the data using the
+    /// coroutine handle.
+    ///
+    /// After the initial snapshot has been produced it returns the name of the created slot and
+    /// the LSN at which we should start the replication stream at.
+    async fn produce_snapshot(
+        &mut self,
+        client: &mut ReplicationClient,
+        col_types: &[PgColumn],
+        timestamper: &Timestamper,
+        co: &mut Co<StreamItem>,
+    ) -> (String, PgLsn) {
+        // TODO: find better naming scheme. Postgres doesn't like dashes
+        let slot_name = Uuid::new_v4().to_string().replace('-', "");
+
+        // Start a transaction and immediatelly create a replication slot with the USE SNAPSHOT
+        // directive. This makes the starting point of the slot and the snapshot of the transaction
+        // identical.
+        //
+        // TODO: The slot should be permanent and destroyed on Drop. This way we can support
+        // reconnects
+        client
+            .simple_query("BEGIN READ ONLY ISOLATION LEVEL REPEATABLE READ;")
+            .await
+            .unwrap();
+        let slot = client
+            .create_logical_replication_slot(
+                &slot_name,
+                true,
+                &self.plugin,
+                Some(SnapshotMode::UseSnapshot),
+            )
+            .await
+            .unwrap();
+
+        let cur_lsn = slot.consistent_point();
+
+        // Initial data copy
+        let copy_query = format!(
+            r#"COPY "{}"."{}" TO STDOUT WITH (FORMAT binary);"#,
+            &self.connector.namespace, &self.connector.table
+        );
+
+        let stream = client.copy_out(&copy_query).await.unwrap();
+
+        let scalar_types = col_types
+            .iter()
+            .map(|c| c.scalar_type.clone())
+            .collect_vec();
+        let rows = BinaryCopyOutStream::new(stream, &scalar_types);
+        pin_mut!(rows);
+
+        let snapshot_ts = timestamper.lease().await;
+
+        // TODO: handle Err case
+        while let Some(Ok(row)) = rows.next().await {
+            self.packer.clear();
+            for (i, ty) in col_types.iter().enumerate() {
+                let datum: Datum = match ty.scalar_type {
+                    PgType::BOOL => row.get::<Option<bool>>(i).into(),
+                    PgType::INT4 => row.get::<Option<i32>>(i).into(),
+                    PgType::INT8 => row.get::<Option<i64>>(i).into(),
+                    PgType::FLOAT4 => row.get::<Option<f32>>(i).into(),
+                    PgType::FLOAT8 => row.get::<Option<f64>>(i).into(),
+                    PgType::DATE => row.get::<Option<NaiveDate>>(i).into(),
+                    PgType::TIME => row.get::<Option<NaiveTime>>(i).into(),
```

Comment:
```TIME``` does not seem to work for me. With this table:

```
CREATE TABLE types_table (date_col DATE, time_col TIME, timestamp_col TIMESTAMP, uuid_col UUID);
INSERT INTO types_table VALUES ('2011-11-11', '11:11:11', '2011-11-11 11:11:11', 'A0EEBC99-9C0B-4EF8-BB6D-6BB9BD380A11');
```

and this ```CREATE SOURCE```:

```
CREATE MATERIALIZED SOURCE "types_table"
  FROM POSTGRES HOST 'host=postgres port=5432 user=postgres password=postgres dbname=postgres'
  PUBLICATION 'mz_source' NAMESPACE 'public' TABLE 'types_table' (date_col DATE, time_col TIME, timestamp_col TIMESTAMP, uuid_col UUID);```

I get:

```
error: unable to handle SQL type: Time
```

Reply:
this seems to work for me. Did you set the replica identity to FULL?

Reply:
This was a bug in testdrive, will submit a PR .